File: top_interview_150/796_rotate_string/main.py
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:

        def shift(s: str) -> str:
            return s[1:] + s[0]

        if s == goal:
            return True

        elif (len(s) != len(goal)) or (set(s) != set(goal)):
            return False

        else:
            shifted = s

        for _ in range(len(s)):

            shifted = shift(shifted)

            if shifted == goal:
                return True

        return False

        

File: top_interview_150/796_rotate_string/main.go
package main

import (
	"fmt"
)

func shift(s []rune) []rune {
	return append(s[1:], s[0])
}


func rotateString(s string, goal string) bool {
    if s == goal {
		return true
	}
	if len(s) != len(goal) {
		return false
	}
	shifted := []rune(s)

	for i := 0; i < len(s); i++ {
		shifted = shift(shifted)
		if string(shifted) == goal {
			return true
		}
	}
	return false
}

func main(){
	s := "abcd"
	goal := "cdab"

	fmt.Printf("Check True: %t\n", rotateString(s,goal))

	s = "bcd"
	goal = "cab"

	fmt.Printf("Check False: %t\n", rotateString(s,goal))
}